### PR TITLE
Implement GetItem function

### PR DESCRIPTION
**Pull request by wkentaro**

### CODE REVIEW

File: tests/chainer_tests/functions_tests/array_tests/test_crop.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    @attr.gpu
+    def test_forward_gpu(self):
+        self.check_forward(cuda.to_gpu(self.x0_data),
+                           cuda.to_gpu(self.x1_data))
+
+    def check_backward(self, x0_data, x1_data, y_grad):
+        x0 = chainer.Variable(x0_data)
+        x1 = chainer.Variable(x1_data)
+        y = functions.crop(x0, x1, self.axes)
+        y.grad = y_grad
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_crop.py]: How about using `gradient_check.check_backward`?
See also https://github.com/pfnet/chainer/blob/ed8ef4786be325899119c3827febc9286da7e354/tests/chainer_tests/functions_tests/noise_tests/test_gaussian.py#L21

File: chainer/functions/array/crop.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            gx = gx.swapaxes(axis, 0)
+            size = ref_ary.shape[axis]
+            start = self.offset
+            end = start + size
+            gx = gx[start:end].swapaxes(axis, 0)
+        gx[...] = gy
+        return gx_orig, None
+
+
+def crop(x0, x1, axes, offset=0):
```
Comment (okuta) [inline chainer/functions/array/crop.py]: How about passing `x1.shape` to `crop`?
The data of `x1` does not used.

File: chainer/functions/array/crop.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        self.axes = axes
+        self.offset = offset
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 2)
+        for axis in self.axes:
+            type_check.expect(in_types[0].ndim > axis)
+            type_check.expect(in_types[1].ndim > axis)
+
+    def forward(self, xs):
```
Comment (okuta) [inline chainer/functions/array/crop.py]: How about following code.
```python
    def forward(self, xs):
        ary, ref_ary = xs
        slices = [slice(None)] * ary.ndim
        for axis in self.axes:
            slices[axis] = slice(self.offset, self.offset + ref_ary.shape[axis])
        return ary[tuple(slices)],
```

File: chainer/functions/array/crop.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        ary, ref_ary = xs
+        for axis in self.axes:
+            ary = ary.swapaxes(axis, 0)
+            size = ref_ary.shape[axis]
+            start = self.offset
+            end = start + size
+            ary = ary[start:end].swapaxes(axis, 0)
+        return ary,
+
+    def backward(self, xs, gys):
```
Comment (okuta) [inline chainer/functions/array/crop.py]: How about following code.
```python
    def backward(self, xs, gys):
        xp = cuda.get_array_module(*xs)
        ref_ary = xs[1]
        gy = gys[0]
        slices = [slice(None)] * gy.ndim
        for axis in self.axes:
            slices[axis] = slice(self.offset, self.offset + ref_ary.shape[axis])
        gx = xp.zeros_like(xs[0])
        gx[tuple(slices)] = gy
        return gx, None
```

**Comment by wkentaro** [comment]
Thanks. I updated the commits.

File: chainer/functions/array/crop.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
+import collections
+
+from chainer import cuda
+from chainer import function
+from chainer.utils import type_check
+
+
+class Crop(function.Function):
+
+    """Function that crops one array with the other with axes and offset."""
```
Comment (okuta) [inline chainer/functions/array/crop.py]: Please fix comment.

File: chainer/functions/array/crop.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        slices = [slice(None)] * gy.ndim
+        for axis in self.axes:
+            slices[axis] = slice(self.offset, self.offset + self.shape[axis])
+        gx = xp.zeros_like(xs[0])
+        gx[tuple(slices)] = gy
+        return gx,
+
+
+def crop(x, shape, axes, offset=0):
+    """Crop one array with the other with axes and offset.
```
Comment (okuta) [inline chainer/functions/array/crop.py]: Please fix comment.

File: chainer/functions/array/crop.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    Args:
+        x (tuple of Variables): Variable to be cropped.
+        shape (tuple or list of ints): Reference shape for cropping.
+            The size in each axis specified by 'axes' is used.
+        axes (int or tuple of ints):
+            Axes that the input array is cropped along with.
+        offset (int): Offset for the starting point of cropping.
+
+    Returns:
+        Variable: :class:``~chainer.Variable`` object
```
Comment (okuta) [inline chainer/functions/array/crop.py]: Please use single back quote.
For example `` :class:`~chainer.Variable` ``.

File: tests/chainer_tests/functions_tests/array_tests/test_crop.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+from chainer import cuda
+from chainer import functions
+from chainer import gradient_check
+from chainer import testing
+from chainer.testing import attr
+
+
+class TestCrop(unittest.TestCase):
+    axes = [1, 2]
+    offset = 0
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_crop.py]: Please add test `offset!=1`.

File: tests/chainer_tests/functions_tests/array_tests/test_crop.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+import chainer
+from chainer import cuda
+from chainer import functions
+from chainer import gradient_check
+from chainer import testing
+from chainer.testing import attr
+
+
+class TestCrop(unittest.TestCase):
+    axes = [1, 2]
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_crop.py]: Please add test `type(axes) == int`.

**Comment by okuta** [comment]
OK. This PR is good except comments.
Please fix `chainer/functions/__init__.py` and `docs/source/reference/functions.rst`.

See also https://github.com/pfnet/chainer/pull/1126/files

**Comment by wkentaro** [comment]
> Please fix chainer/functions/__init__.py

What do I missing? Is this not enough?
screen shot 2016-05-23 at 00 43 58

**Comment by wkentaro** [comment]
Actually, I found the `offset` argument should be the list of ints for each axis.
Sorry for this because the review is almost finished.

File: chainer/functions/array/crop.py (hunk 5 above)
Comment (wkentaro) [inline chainer/functions/array/crop.py]: Fixed.

File: chainer/functions/array/crop.py (hunk 6 above)
Comment (wkentaro) [inline chainer/functions/array/crop.py]: Fixed.

File: chainer/functions/array/crop.py (hunk 7 above)
Comment (wkentaro) [inline chainer/functions/array/crop.py]: Fixed.

File: tests/chainer_tests/functions_tests/array_tests/test_crop.py (hunk 8 above)
Comment (wkentaro) [inline tests/chainer_tests/functions_tests/array_tests/test_crop.py]: Fixed.

File: tests/chainer_tests/functions_tests/array_tests/test_crop.py (hunk 9 above)
Comment (wkentaro) [inline tests/chainer_tests/functions_tests/array_tests/test_crop.py]: Fixed.

**Comment by wkentaro** [comment]
> Please fix  docs/source/reference/functions.rst.

Fixed.

**Comment by wkentaro** [comment]
I think I have fixed against all of your comments. Please review again when you have time.

**Comment by beam2d** [comment]
Thank you for proposing/implementing this functionality. I read the PR, and felt that it is much better to support cropping as a special case of differentiable basic indexing. The current implementation can be easily modified to support arbitrary slicing (since both NumPy and CuPy supports basic indexing), and then the interface is much more compatible to NumPy/CuPy. You can set `Variable.__getitem__` to add an indexing operator of Variable. See `chainer/functions/math/basic_math.py` for an example of injecting such an operator. If you can work on it, the Function would be renamed to `GetItem`, and the module name etc. as well.

**Comment by wkentaro** [comment]
Updated. Is this ok?

**Comment by okuta** [comment]
Very good!
Please add test about `slices = slice(2, 5)`.
https://github.com/pfnet/chainer/pull/1154/files#diff-278878af6cd00cabfd7437aa9efff4f2R71

**Comment by wkentaro** [comment]
Updated.

File: chainer/functions/array/get_item.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        gx[tuple(self.slices)] = gy
+        return gx,
+
+
+def get_item(x, slices):
+    """Extract elements from array with specified shape, axes and offsets.
+
+    Args:
+        x (tuple of Variables): Variable to be sliced.
+        slices (tuple of slice objects): Values to slice variable in each axis.
```
Comment (okuta) [inline chainer/functions/array/get_item.py]: Please fix this text.

File: chainer/functions/array/get_item.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+    Returns:
+        Variable: :class:`~chainer.Variable` object
+            which is cropped array of ``x``.
```
Comment (okuta) [inline chainer/functions/array/get_item.py]: Please fix this text.

File: docs/source/reference/functions.rst [hunk 12]
```diff
@@ -99,6 +99,10 @@ copy
 ~~~~
 .. autofunction:: copy
 
+crop
+~~~~
+.. autofunction:: crop
```
Comment (okuta) [inline docs/source/reference/functions.rst]: Please fix this statement.

**Comment by okuta** [comment]
LGTM except documents.

**Comment by wkentaro** [comment]
Fixed.

**Comment by okuta** [comment]
Thank you very much!

**Comment by wkentaro** [comment]
Thank you for your reviews!
@okuta @beam2d

2016年5月24日火曜日、Ryosuke Okuta<notifications@github.com>さんは書きました:

> Thank you very much!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/pfnet/chainer/pull/1154#issuecomment-221202464>
>


-- 
和田 健太郎 / Kentaro Wada
http://wkentaro.com

**Pull request merged**